README
General
=======

This was a 'quick' project to make a notifier script to email me (actual command is configurable) when certain interesting events happen with my
deviantART account (www.deviantart.com - currently comments, replies, unread notes and watched people's deviations posted), optionally with a whitelist
to only bug me when certain people are involved for certain event types, checking every 5 minutes by default.

I don't have a need for further functionality, but as usual if there is enough demand, I can extend it. Please star this project if you use it so
that I know which projects are most popular and therefore worth working on.

The deviantAnywhere Firefox addon (https://addons.mozilla.org/en-US/firefox/addon/deviantanywhere/) was used as an example when developing the
deviantART service code. 


Requirements
============

Tested with (if you get it working on earlier versions please say, all available in Debian repos):

Python v3.4+
Beautiful Soup v4.3.2: http://www.crummy.com/software/BeautifulSoup/
lxml v3.6.4 (used as the HTML parser with Beautiful Soup): http://lxml.de/
Requests v2.3.0: http://python-requests.org
python-YAML v3.11: http://pyyaml.org/


Documentation
=============

This is a very simple script currently - it takes no arguments and runs in a loop, polling your deviantART account after a delay. To configure,
create the '~/.config/deviantart-checker' directory and copy/rename example.conf to 'deviantart-checker.conf' inside, editing it as needed (see the
comments in the file).

If you follow the sendemail example, when someone replies to a comment you made, you'll get an email like the following:

---------------------------------------------------------------------------

Subject: [deviantart-checker] New Replies

Message:

New Replies:

On <deviation title> by <artist>:

<user> posted:
blah blah blah blah

---------------------------------------------------------------------------

The output has also been tailored for other events and is similar.


Contact Details
===============

OmegaPhil: OmegaPhil@startmail.com

NEWS
deviantart-deviations-downloader.py v0.2 and updated devart.py (16.02.17):
deviantART has reworked their HTMl a little meaning that all scripts failed
to log in, and then once that was fixed, deviation downloading failed completely
as the gallery widget had been completely redone. This has now been sorted, with
the exception that

*deviation folders are no longer savable*,

as the gallery widget no longer contains that information. It is available via a
separate pass to analyse the folders specially, but I don't need this currently
so it hasn't been implemented.

deviantart-checker.py v1 (18.10.16):
deviantartART interaction code has now been moved out into devart.py.
BREAKING CHANGE: Credentials are now stored in
'~/.config/deviantart-scripts/credentials.conf' for clean sharing with other
scripts in the (now) suite.
BREAKING CHANGE: Configuration is now read from the ''~/.config/deviantart-scripts'
directory - please move your configuration manually.
Note content is now fetched when reporting on unread notes (courtesy of
progression with the notes downloader script).

deviantart-notes-downloader.py v0.1 (18.10.16):
Script created to allow backing up sent and received notes.

deviantart-unread-sent-notes-checker.py v0.1 (18.10.16):
Script created to allow monitoring when sent notes become read.

deviantart-deviations-downloader.py v0.1 (18.10.16):
Script created to allow backing up gallery structure and deviation detail.

v0.5 (21.03.15):
Only allow one instance of the script to run at once.

v0.4 (11.10.14):
command_to_run_on_failure in the config now allows you to run a configurable
command when an unhandled error happens. The script continues to run under these
circumstances, but now you don't have to wrap around stderr somehow to get told
when it happens.

v0.3 (08.09.14):
Allow for the whitelist to be selectively applied to certain events only - e.g.
filter deviations but always be notified on replies.
Improve text generation from initially HTML information associated with events. 

v0.2 (30.08.14):
Initial public release.


deviantart-unread-sent-notes-checker.py
#!/usr/bin/env python3

'''
Version 0.1 2016.10.18
Copyright (c) 2016, OmegaPhil - OmegaPhil@startmail.com

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

import datetime
import fcntl
import io
import numbers
import os
import os.path
import subprocess
import time
import traceback
import shlex
import sys

import yaml

import devart


config = {}


def generate_command_fragments(command, subject, message):
    '''Prepare command with variable substitution'''

    # Substituting values into command to call in the normal way for
    # the subject since that is essentially fixed - content is not
    # fixed and can contain quotes and speechmarks. I have tried to
    # escape such stuff with shlex.quote, but it can't cope in
    # examples of single quotes, and when a full command is quoted
    # it doesn't execute correctly. Given that I don't want this going
    # through a shell, I should be able to give command parameters
    # straight to the called process without weird escaping - which is
    # what I am doing here
    command = command.replace('%s', '[deviantart-unread-sent-notes-checker] '
                              + subject)
    return [fragment.replace('%m', message) for fragment in
            shlex.split(command)]


def handle_unknown_error(err):
    '''Handle unknown error'''

    # pylint: disable=line-too-long
    # I have decided to just log failures rather than die - this is
    # basically a service, and it is not acceptable for temporary
    # failures to stop the program doing its job - it must simply retry
    # after the usual delay
    # Examples of caught errors:
    # ConnectionError: HTTPSConnectionPool(host='www.deviantart.com', port=443): Max retries exceeded with url: /users/login (Caused by <class 'socket.error'>: [Errno 113] No route to host)
    # (<urllib3.connection.VerifiedHTTPSConnection object at 0x7f7f5f3a5e50>, 'Connection to www.deviantart.com timed out. (connect timeout=60)')
    error_message = 'Unhandled error \'%s\'\n\n%s' % (err, traceback.format_exc())
    print(error_message, file=sys.stderr)

    # Running command_to_run_on_failure if specified
    if config['command_to_run_on_failure']:

        try:
            command_fragments = generate_command_fragments(config['command_to_run_on_failure'],
                                                           'Error', error_message)

            # Running command without a shell
            subprocess.call(command_fragments)

        except Exception as e:  # pylint: disable=broad-except
            print('Calling the command to run on failure failed:\n\n%s\n'
                  '\n%s\n' % (config['command_to_run_on_failure'], e),
                  file=sys.stderr)


def load_config():
    '''Load config'''

    global config  # pylint: disable=global-statement

    # Loading configuration if it exists. Credentials has been split out into
    # its own file so that it can be shared amongst various scripts
    config_directory = os.path.expanduser('~/.config/deviantart-scripts')
    config_file_path = os.path.join(config_directory,
                                    'deviantart-unread-sent-notes-checker.conf')
    credentials_file_path = os.path.join(config_directory, 'credentials.conf')
    if (os.path.exists(config_file_path)
        and os.path.exists(credentials_file_path)):

        # Loading YAML documents - theres no need for them to be genuinely
        # separate documents, so just sticking them together
        try:
            config_text = (io.open(config_file_path, 'r').read() + '\n' +
                            io.open(credentials_file_path, 'r').read())
            config = yaml.load(config_text, yaml.CLoader)
            if config is None:
                raise Exception('YAML documents empty')
        except Exception as e:
            raise Exception('Unable to load config from YAML documents '
                            '\'%s\' and \'%s\':\n\n%s\n\n%s\n'
                            % (config_file_path, credentials_file_path, str(e),
                               traceback.format_exc()))

    # Ensuring required settings exist
    if 'username' not in config or 'password' not in config:
        raise Exception('Please ensure a deviantART username and password is '
                        'configured in \'%s\'' % credentials_file_path)
    if 'command_to_run' not in config:
        raise Exception('Please ensure command_to_run is configured in \'%s\'' %
                        config_file_path)

    # Ensuring sensible defaults
    # Keep update delay at at least 5 minutes so as not to cause deviantART
    # unnecessary load
    if ('update_every_minutes' not in config or
            not isinstance(config['update_every_minutes'], numbers.Number) or
            not config['update_every_minutes'] >= 5):
        config['update_every_minutes'] = 5


def poll_service():
    '''Main loop'''

    # Logging in to deviantART - any errors here will be fatal and are handled
    # in the main scope
    dA = devart.DeviantArtService(config['username'], config['password'])

    # Looping for regular unread notes fetching
    current_unread_notes = []
    while True:

        try:

            # Attempting to log in - errors at this level will be logged and the
            # program will simply wait until the next interval to try again
            if not dA.logged_in:
                dA.login()

            try:

                # Getting the current unread sent notes
                latest_unread_notes = dA.get_unread_sent_notes()

            # Currently I'll treat all exceptions here as issues with
            # deviantART or an expired login - invalidate the login and report
            # on the issue but reraise waiting for the next loop run to react
            except Exception as e:
                dA.logged_in = False
                raise

            # Determining newly-read notes
            read_notes_change_summary = []
            read_notes = set(current_unread_notes) - set(latest_unread_notes)
            current_unread_notes = latest_unread_notes
            if read_notes:
                read_notes_change_summary.append('The following sent notes have'
                                                 ' now been read:')
                for read_note in read_notes:
                    timestamp = datetime.datetime.fromtimestamp(read_note.ts)
                    # timestamp = timestamp.strftime('%y/%m/%d %H:%M:%S')
                    note_details = ('\'%s\' sent to %s on %s'
                    % (read_note.title, read_note.recipient, timestamp))
                    read_notes_change_summary.append(note_details)

                try:
                    command_fragments = generate_command_fragments(
                        config['command_to_run'], 'Freshly-Read Notes',
                        "\n\n".join(read_notes_change_summary))

                    # Running command without a shell
                    subprocess.call(command_fragments)

                except Exception as e:  # pylint: disable=broad-except
                    raise Exception('Calling the command to run to notify read '
                                    'notes failed:\n\n%s\n\n%s\n' %
                                    (config['command_to_run'], e))

        except Exception as e:  # pylint: disable=broad-except
            handle_unknown_error(e)

        time.sleep(config['update_every_minutes'] * 60)


# Loading config
try:
    load_config()
except Exception as e:  # pylint: disable=broad-except
    print('Unable to load or invalid configuration file:\n\n%s' % e,
          file=sys.stderr)
    sys.exit(1)

with open(__file__) as f:
    try:

        # Only allow one instance at a time - exclusive lock on the script
        # itself rather than using a pidfile
        fcntl.flock(f, fcntl.LOCK_EX | fcntl.LOCK_NB)

        # Start polling loop
        poll_service()

    except IOError as e:
        print('This script appears to already be running - please kill all '
              'other instances before running again', file=sys.stderr)
        sys.exit(1)

    except Exception as e:  # pylint: disable=broad-except
        handle_unknown_error(e)

    finally:

        # Release lock
        fcntl.flock(f, fcntl.LOCK_UN | fcntl.LOCK_NB)


deviantart-checker.py
#!/usr/bin/env python3

'''
Version 1 2016.10.18
Copyright (c) 2014-2016, OmegaPhil - OmegaPhil@startmail.com

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

import fcntl
import io
import numbers
import os
import os.path
import subprocess
import time
import traceback
import shlex
import sys

import yaml

import devart


config = {}


def generate_command_fragments(command, subject, message):
    '''Prepare command with variable substitution'''

    # Substituting values into command to call in the normal way for
    # the subject since that is essentially fixed - content is not
    # fixed and can contain quotes and speechmarks. I have tried to
    # escape such stuff with shlex.quote, but it can't cope in
    # examples of single quotes, and when a full command is quoted
    # it doesn't execute correctly. Given that I don't want this going
    # through a shell, I should be able to give command parameters
    # straight to the called process without weird escaping - which is
    # what I am doing here
    command = command.replace('%s', '[deviantart-checker] ' + subject)
    return [fragment.replace('%m', message) for fragment in
            shlex.split(command)]


def handle_unknown_error(err):
    '''Handle unknown error'''

    # pylint: disable=line-too-long
    # I have decided to just log failures rather than die - this is
    # basically a service, and it is not acceptable for temporary
    # failures to stop the program doing its job - it must simply retry
    # after the usual delay
    # Examples of caught errors:
    # ConnectionError: HTTPSConnectionPool(host='www.deviantart.com', port=443): Max retries exceeded with url: /users/login (Caused by <class 'socket.error'>: [Errno 113] No route to host)
    # (<urllib3.connection.VerifiedHTTPSConnection object at 0x7f7f5f3a5e50>, 'Connection to www.deviantart.com timed out. (connect timeout=60)')
    error_message = 'Unhandled error \'%s\'\n\n%s' % (err, traceback.format_exc())
    print(error_message, file=sys.stderr)

    # Running command_to_run_on_failure if specified
    if config['command_to_run_on_failure']:

        try:
            command_fragments = generate_command_fragments(config['command_to_run_on_failure'],
                                                           'Error', error_message)

            # Running command without a shell
            subprocess.call(command_fragments)

        except Exception as e:  # pylint: disable=broad-except
            print('Calling the command to run on failure failed:\n\n%s\n'
                  '\n%s\n' % (config['command_to_run_on_failure'], e),
                  file=sys.stderr)


def load_config():
    '''Load config'''

    global config  # pylint: disable=global-statement

    # Loading configuration if it exists. Credentials has been split out into
    # its own file so that it can be shared amongst various scripts
    config_directory = os.path.expanduser('~/.config/deviantart-scripts')
    config_file_path = os.path.join(config_directory, 'deviantart-checker.conf')
    credentials_file_path = os.path.join(config_directory, 'credentials.conf')
    if (os.path.exists(config_file_path)
        and os.path.exists(credentials_file_path)):

        # Loading YAML documents - theres no need for them to be genuinely
        # separate documents, so just sticking them together
        try:
            config_text = (io.open(config_file_path, 'r').read() + '\n' +
                            io.open(credentials_file_path, 'r').read())
            config = yaml.load(config_text, yaml.CLoader)
            if config is None:
                raise Exception('YAML documents empty')
        except Exception as e:
            raise Exception('Unable to load config from YAML documents '
                            '\'%s\' and \'%s\':\n\n%s\n\n%s\n'
                            % (config_file_path, credentials_file_path, str(e),
                               traceback.format_exc()))

    # Ensuring required settings exist
    if 'username' not in config or 'password' not in config:
        raise Exception('Please ensure a deviantART username and password is '
                        'configured in \'%s\'' % credentials_file_path)
    if 'command_to_run' not in config:
        raise Exception('Please ensure command_to_run is configured in \'%s\'' %
                        config_file_path)

    # Ensuring sensible defaults
    # Keep update delay at at least 5 minutes so as not to cause deviantART
    # unnecessary load
    if ('update_every_minutes' not in config or
            not isinstance(config['update_every_minutes'], numbers.Number) or
            not config['update_every_minutes'] >= 5):
        config['update_every_minutes'] = 5

    # Can't get the indentation to pylint's liking
    # pylint: disable=bad-continuation
    # Validating apply_whitelist_to
    if (('apply_whitelist_to' not in config or
         not config['apply_whitelist_to']
        ) and
        'notification_whitelist' in config and
        config['notification_whitelist']):
        print('config specifies notification_whitelist however '
              'apply_whitelist_to is not present or empty - whitelist will not '
              'be used', file=sys.stderr)
    for event in config['apply_whitelist_to']:
        if event not in ['comments', 'replies', 'unread_notes', 'deviations']:
            print('\'%s\' in \'apply_whitelist_to\' configuration is invalid -'
                  ' please use \'comments\'/\'replies\'/\'unread_notes\''
                  '/\'deviations\'' % event, file=sys.stderr)


def poll_service():
    '''Main loop'''

    # pylint: disable=too-many-locals,too-many-branches,too-many-statements

    # Logging in to deviantART - any errors here will be fatal and are handled
    # in the main scope
    dA = devart.DeviantArtService(config['username'], config['password'])
    state = devart.AccountState('~/.cache/deviantart-scripts/deviantart-checker-state.txt')

    # Looping for regular message fetching
    while True:

        try:

            # Attempting to log in - errors at this level will be logged and the
            # program will simply wait until the next interval to try again
            if not dA.logged_in:
                dA.login()

            try:

                # Getting the current state of messages
                dA.get_messages(state)

            # Currently I'll treat all exceptions here as issues with
            # deviantART or an expired login - invalidate the login and report
            # on the issue but reraise waiting for the next loop run to react
            except Exception as e:
                dA.logged_in = False
                raise

            # Working out how the state has changed
            new_comments = devart.get_new(state, devart.COMMENTS)
            new_replies = devart.get_new(state, devart.REPLIES)
            new_unread_notes = devart.get_new(state, devart.UNREAD_NOTES)
            new_deviations = devart.get_new(state, devart.DEVIATIONS)

            # Setting default change reporting state based on whether there is
            # a notification whitelist in use, and then the particular events
            # affected
            notification_whitelist_comments = False
            notification_whitelist_replies = False
            notification_whitelist_unread_notes = False
            notification_whitelist_deviations = False
            if ('notification_whitelist' in config and
                    config['notification_whitelist'] and
                    'apply_whitelist_to' in config and
                    config['apply_whitelist_to']):
                if 'comments' in config['apply_whitelist_to']:
                    notification_whitelist_comments = True
                if 'replies' in config['apply_whitelist_to']:
                    notification_whitelist_replies = True
                if 'unread_notes' in config['apply_whitelist_to']:
                    notification_whitelist_unread_notes = True
                if 'deviations' in config['apply_whitelist_to']:
                    notification_whitelist_deviations = True

            # Summarise changes, and when a whitelist is in place, only
            # returning information if it includes something generated from a
            # person of interest
            # Filter is used to remove empty items, list is used to explicitly
            # evaluate immediately
            comments_change_title, comments_change_summary, comments_users = summarise_changes(new_comments, devart.COMMENTS)  # pylint: disable=line-too-long
            if ((notification_whitelist_comments and not set(config['notification_whitelist']).intersection(comments_users)) or  # pylint: disable=line-too-long
                    not comments_change_title):
                comments = ''
            else:
                comments = '%s:\n%s' % (comments_change_title,
                                        comments_change_summary)

            replies_change_title, replies_change_summary, replies_users = summarise_changes(new_replies, devart.REPLIES)  # pylint: disable=line-too-long
            if ((notification_whitelist_replies and not set(config['notification_whitelist']).intersection(replies_users)) or  # pylint: disable=line-too-long
                    not replies_change_title):
                replies = ''
            else:
                replies = '%s:\n%s' % (replies_change_title,
                                       replies_change_summary)

            unread_notes_change_title, unread_notes_change_summary, unread_notes_users = summarise_changes(new_unread_notes, devart.UNREAD_NOTES)  # pylint: disable=line-too-long
            if ((notification_whitelist_unread_notes and not set(config['notification_whitelist']).intersection(unread_notes_users)) or  # pylint: disable=line-too-long
                    not unread_notes_change_title):
                unread_notes = ''
            else:
                unread_notes = '%s:\n%s' % (unread_notes_change_title,
                                            unread_notes_change_summary)

            deviations_change_title, deviations_change_summary, deviations_users = summarise_changes(new_deviations, devart.DEVIATIONS)  # pylint: disable=line-too-long
            if ((notification_whitelist_deviations and not set(config['notification_whitelist']).intersection(deviations_users)) or  # pylint: disable=line-too-long
                    not deviations_change_title):
                deviations = ''
            else:
                deviations = '%s:\n%s' % (deviations_change_title,
                                          deviations_change_summary)

            # pylint: disable=bad-builtin
            title_bits = list(filter(None, [deviations_change_title,
                                            unread_notes_change_title,
                                            replies_change_title,
                                            comments_change_title]))
            content = list(filter(None, [deviations, unread_notes, replies,
                                         comments]))

            if content:

                try:
                    # pylint: disable=line-too-long
                    command_fragments = generate_command_fragments(config['command_to_run'],
                                                                   ", ".join(title_bits),
                                                                   "\n\n".join(content))

                    # Running command without a shell
                    subprocess.call(command_fragments)

                except Exception as e:  # pylint: disable=broad-except
                    raise Exception('Calling the command to run to notify changes '
                                    'failed:\n\n%s\n\n%s\n' %
                                    (config['command_to_run'], e))

        except Exception as e:  # pylint: disable=broad-except
            handle_unknown_error(e)

        time.sleep(config['update_every_minutes'] * 60)


def summarise_changes(messages, messages_type):
    '''A report summarising new messages of the relevant type'''

    # pylint: disable=redefined-outer-name, too-many-branches
    # pylint: disable=too-many-locals

    users = []

    # Returning null-length strings if no actual change has happened
    if not messages:
        return '', '', []

    # Dealing with different message types
    if messages_type == devart.COMMENTS or messages_type == devart.REPLIES:

        # Sorting comments on page they were posted under then the
        # timestamp - body is included so that it can be used later
        new_comments = sorted([(comment.title, comment.ts, comment.who,
                                comment.body)
                               for comment in messages],
                              key=lambda comment: (comment[0], comment[1],
                                                   comment[2], comment[3]))

        # Generating and returning summary
        current_title = None
        summary = []
        for title, _, who, body in new_comments:
            if current_title != title:
                summary.append('\nOn ' + title + ':\n')
                current_title = title
            summary.append('%s posted:\n%s' % (who, body))

            # Keeping record of users causing the updates
            if not who in users:
                users.append(who)

        if messages_type == devart.COMMENTS:
            messages_type_text = 'Comments'
        else:
            messages_type_text = 'Replies'
        return 'New ' + messages_type_text, '\n'.join(summary), users

    elif messages_type == devart.UNREAD_NOTES:

        # Sorting unread notes based on sender then title
        new_unread_notes = sorted([(note.sender, note.title, note.text)
                                   for note in messages],
                                  key=lambda note: (note[0], note[1]))

        # Generating and returning summary, now including note text
        current_sender = None
        summary = []
        for sender, title, text in new_unread_notes:
            if current_sender != sender:
                summary.append('\n' + sender + ' sent:\n')
                current_sender = sender
            summary += [title, '=' * len(title), text, '\n']

            # Keeping record of users causing the updates
            if not sender in users:
                users.append(sender)

        return 'New Unread Notes', '\n'.join(summary), users

    elif messages_type == devart.DEVIATIONS:

        # Generating case-insensitive sorted list of usernames and titles,
        # in that order
        new_deviations = sorted([(deviation.username, deviation.title)
                                 for deviation in messages],
                                key=lambda deviation: (deviation[0].lower(),
                                                       deviation[1].lower()))  # pylint: disable=line-too-long

        # Generating and returning summary
        current_username = None
        summary = []
        for username, title in new_deviations:
            if current_username != username:
                summary.append('\n' + username + ':\n')
                current_username = username
            summary.append(title)

            # Keeping record of users causing the updates
            if not username in users:
                users.append(username)

        return 'New Deviations', '\n'.join(summary), users

    else:

        # Invalid messages_type passed
        raise Exception('summarise_changes was called with an invalid '
                        'messages_type (%s)' % messages_type)


# Loading config
try:
    load_config()
except Exception as e:  # pylint: disable=broad-except
    print('Unable to load or invalid configuration file:\n\n%s' % e,
          file=sys.stderr)
    sys.exit(1)

with open(__file__) as f:
    try:

        # Only allow one instance at a time - exclusive lock on the script
        # itself rather than using a pidfile
        fcntl.flock(f, fcntl.LOCK_EX | fcntl.LOCK_NB)

        # Start polling loop
        poll_service()

    except IOError as e:
        print('This script appears to already be running - please kill all '
              'other instances before running again', file=sys.stderr)
        sys.exit(1)

    except Exception as e:  # pylint: disable=broad-except
        handle_unknown_error(e)

    finally:

        # Release lock
        fcntl.flock(f, fcntl.LOCK_UN | fcntl.LOCK_NB)
